core/src/messages.cpp

```diff
@@ -462,28 +486,71 @@ void from_json(const nlohmann::json& doc, curtain_query& query) noexcept (false)
 
     const auto& args = doc.at("args");
 
-    std::vector< std::vector< int > > coords;
-    args.at("coords").get_to(coords);
-    query.dim0s.reserve(coords.size());
-    query.dim1s.reserve(coords.size());
-
-    try {
-        for (const auto& pair : coords) {
-            query.dim0s.push_back(pair.at(0));
-            query.dim1s.push_back(pair.at(1));
+    auto extract_coords = [&]<typename T>(auto mapping, T) {
+        std::vector< std::vector< T > > coords;
+        args.at("coords").get_to(coords);
+        query.dim0s.reserve(coords.size());
+        query.dim1s.reserve(coords.size());
+
+        try {
+            for (const auto& pair : coords) {
+                const auto dims = mapping(pair.at(0), pair.at(1));
+                query.dim0s.push_back(dims.first);
+                query.dim1s.push_back(dims.second);
+            }
+        } catch (std::out_of_range&) {
+            throw bad_value("bad coord arg; expected list-of-pairs");
         }
-    } catch (std::out_of_range&) {
-        throw bad_value("bad coord arg; expected list-of-pairs");
-    }
+    };
 
     const std::string& kind = args.at("kind");
     if (kind == "index") {
         /* no-op - already cartesian indices */
+        extract_coords(
+            [](int x, int y){return std::pair{x, y};},
+            int()
+        );
     }
     else if (kind == "lineno") {
-        const auto& line_numbers = query.manifest.line_numbers;
-        to_cartesian_inplace(line_numbers[0], query.dim0s);
-        to_cartesian_inplace(line_numbers[1], query.dim1s);
+        extract_coords(
+            [&](int x, int y) {
+                const auto &line_numbers = query.manifest.line_numbers;
+                return std::pair(
+                    to_cartesian(line_numbers[0], x),
+                    to_cartesian(line_numbers[1], y)
+                );
+            },
+            int()
+        );
+    }
+    else if (kind == "utm") {
```

Comment:
It would require fetching several of these blocks from storage. In addition many of them will contain 0 values in the case where there is no trace for the selected point. Users can get the actual cdp values that were read from the file when retrieving the result. This way they will know the actual position of the traces they get back int case the points did not hit the traces perfectly.